### PR TITLE
fix(code-intel): isolate maximum OOM preference to indexer children

### PR DESCRIPTION
**Pull request by WingedGuardian**
## Why

A resource cap bounds how much a batch index may use; it does not decide which eligible process the kernel selects when the containing cgroup runs out of memory. Code-intelligence indexing is disposable and repeatable, while its supervisor and interactive sessions carry control state or unsaved work.

## What changed

- The disposable indexer child writes and verifies oom_score_adj=1000 immediately before exec.
- The launcher, systemd-run process, pressure watchdog, and capability probe retain their inherited score so they remain able to pause, kill, and reap the workload.
- The same child-only behavior applies to the user-systemd scope path and the nice/ionice/rlimit fallback.
- The heavy workload refuses to start with exit 125 when the maximum adjustment cannot be established and read back.
- The operator setting may be empty or exactly 1000; lower, negative, malformed, and oversized values fail closed rather than weakening the safety contract.
- The script canonicalizes its own entrypoint before GitNexus changes working directory, while preserving argv and child exit status.

The maximum preference is still scoped by kernel OOM eligibility and cgroup domain. This launcher does not reprioritize workers already running in an external daemon, does not make OOM impossible, and does not replace admission control or memory caps.

## Why 1000

Genesis CC subprocesses use a 500 rung. A lower or equal indexer value cannot guarantee that a disposable batch child outranks an interactive subprocess under the same OOM domain. Linux documents 1000 as the maximum adjustment, so the batch child uses that value and no weaker override is accepted.

## Verification

- Adversarial review: no P1/P2 findings; Ready to merge: Yes; 97% confidence before the two remaining P3 test-proof gaps were closed.
- 98 affected tests pass: 71 index-entrypoint tests plus 27 runner/freeze tests.
- Deterministic tests cover both systemd and fallback paths, actual execution-invocation score isolation, descendant process-group cleanup, adjustment failure, argv boundaries, exit propagation, worktree no-op behavior, and relative entrypoint handling.
- Private runtime E2E with real user systemd and the real kernel: supervisor adjustment 0, child adjustment 1000, child inside the named scope, exit 23 propagated, scope inactive afterward, and cgroup oom_kill unchanged at 54.
- bash -n, Ruff, git diff --check, and ShellCheck pass; ShellCheck reports only pre-existing informational diagnostics.
- Genesis health after E2E: healthy, 36 GiB container limit, 17 GiB available, zero memory PSI.

This PR does not run either real indexer and does not alter the GitNexus freeze or Codebase MCP disable state.

## Summary by CodeRabbit

* **New Features**
  * Code intelligence indexing jobs now receive a default out-of-memory priority setting.
  * Valid environment overrides can customize the priority, while higher inherited values are preserved.
  * Indexing refuses to start when the configured value is invalid or cannot be safely applied.
  * The setting applies consistently across supported scheduling methods, without changing supervisor behavior.
* **Documentation**
  * Updated the documented measured working-set requirement from 4.65 GiB to 4.54 GiB.

E2E: after merge, rerun the private real-systemd child probe and confirm the supervisor keeps its inherited score, the child reads 1000 inside the named scope, exit status propagates, and the scope tears down without an OOM.

### CODE REVIEW

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
Structural review: CLEAN — no pattern violations found.
PII/Secrets/Wording scan: CLEAN

Checked: `scripts/lib/code_intel_index.sh` (new `_apply_oom_score_adj` function — a self-write to `/proc/self/oom_score_adj` for OOM kill-order preference, called before the `systemd-run` probe), `tests/test_scripts/test_code_intel_index.py` (8 new tests covering the default, override, non-numeric, negative, raise-only, explicit-lowers-with-log, oversized/overflow, and cc-subprocess-rung cases), and the changelog entry.

- No `nohup`, direct `task_states` inserts, `/tmp/` usage, `git add .`/`-A`, force-push/reset --hard, bare `except:`, mutable default args, unvalidated `os.killpg`, or raw SQL against `genesis.db` in the diff.
- No API keys/tokens, private IPs, SSH keys, or DB connection strings in the added lines; no personal emails, username-bearing paths, or private hostnames either (commit author email is the standard `@users.noreply.github.com`).
- PR title/body/changelog: no security-incident, PII-leak, or infrastructure-exposure wording — the whole PR is about kernel `oom_score_adj` semantics and validation, described without naming any private infra.
- Commit carries a valid `feat(code-intel):` conventional prefix.
- Bash-only change so Domain 1's Python-specific import-hygiene/docstring checks don't apply; the new shell function is heavily commented and the raise-only/overflow/octal edge cases are each pinned by a dedicated test.

VERDICT: PASS



---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/lib/code_intel_index.sh [hunk 1]
```diff
@@ -146,7 +146,142 @@ else
...
+    # Someone who exports CODE_INTEL_INDEX_OOM_SCORE_ADJ has stated an intent,
+    # and quietly ignoring a lever because it happens to lower the value would
+    # be its own surprise — the lever exists to be obeyed. So: explicit wins and
+    # says what it did; the default defers to a higher inherited value.
+    local want explicit=0
+    if [ -n "${CODE_INTEL_INDEX_OOM_SCORE_ADJ:-}" ]; then
+        want="$CODE_INTEL_INDEX_OOM_SCORE_ADJ"
+        explicit=1
+    else
+        want=900
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/lib/code_intel_index.sh]: Fixed at c3b63ab65: the disposable child now establishes and verifies oom_score_adj=1000 immediately before exec, while the supervisor and watchdog keep their inherited score. Tests and a real-systemd E2E verify the child-only boundary.

File: scripts/lib/code_intel_index.sh [hunk 2]
```diff
@@ -146,7 +146,142 @@ else
...
+        _log "note: CODE_INTEL_INDEX_OOM_SCORE_ADJ=$canonical LOWERS the inherited $current (explicit override wins over raise-only)"
+    fi
+
+    if printf '%s\n' "$canonical" > /proc/self/oom_score_adj 2>/dev/null; then
+        # Deliberately does NOT name other units' scores. An earlier version said
+        # "the server at 100", which no shipped configuration set — the unit
+        # template declared -500 (a value the user manager silently refuses, which
+        # is a separate fix). Quoting a number this script does not own turns an
+        # operational log line into a false claim about kill ordering.
+        _log "oom_score_adj=$canonical — this job is the preferred OOM victim, ahead of CC subprocesses, the server and the session"
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/lib/code_intel_index.sh]: Fixed at c3b63ab65: weaker overrides are no longer supported. Empty or exactly 1000 is accepted; lower, malformed, negative, and oversized values refuse the heavy workload before exec.

File: tests/test_scripts/test_code_intel_index.py [hunk 3]
```diff
@@ -542,3 +553,207 @@ def test_triggers_enqueue_markers_and_do_not_spawn():
...
+    plus inheritance. What has to hold is that the tool actually sees it.
+    """
+    fakebin, log = tmp_path / "fakebin", tmp_path / "tools.log"
+    slog = tmp_path / "systemd-run.log"
+    _fake_tools(fakebin, log)
+    _fake_systemd_run(fakebin, slog, probe_ok=True)
+    repo = _make_repo(tmp_path)
+    res = _run_entry(tmp_path, repo, "cbm", path=f"{fakebin}:{_SYSTEM_PATH}")
+    assert res.returncode == 0, res.stderr
+    assert "OOM_ADJ:900" in log.read_text()  # ABOVE invoker.py's 500 for CC subprocesses
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply tests/test_scripts/test_code_intel_index.py]: Fixed at c3b63ab65: assertions now derive the required score and validate the actual child, including the real execution-time systemd invocation rather than only the capability probe.

File: scripts/lib/code_intel_index.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/lib/code_intel_index.sh]: Fixed at c3b63ab65 using the recommended architecture: oom_score_adj=1000 is applied and read back only in the disposable child immediately before exec. The launcher, systemd-run process, probe, and watchdog retain their inherited score. Full affected suites pass 98/98 and a private real-systemd E2E confirmed supervisor=0, child=1000, correct scope membership, rc propagation, no OOM, and clean scope teardown.

**Comment by WingedGuardian** [comment]
@codex review

File: tests/test_scripts/test_code_intel_index.py [hunk 4]
```diff
@@ -1072,3 +1084,367 @@ def test_fractional_and_malformed_caps(tmp_path):
...
+    # test in the session, and the value is not 0 everywhere: a GitHub runner
+    # starts at 500 (measured), so resetting to 0 would silently change what
+    # _inherited_oom_adj() returns for the cells that assert on it.
+    adj = Path("/proc/self/oom_score_adj")
+    original = adj.read_text().strip()
+    adj.write_text("1000\n")
+    try:
+        res = _run_entry(tmp_path, repo, "cbm", path=f"{fakebin}:{_SYSTEM_PATH}")
+    finally:
+        adj.write_text(f"{original}\n")
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply tests/test_scripts/test_code_intel_index.py]: Fixed in a0e85806e. Both inherited-score setups now run only in the subprocess via preexec_fn; the pytest parent is never modified or restored. The two affected tests passed three consecutive runs, and Ruff/diff checks are clean.
